### ISSUE TITLE
Security, WoT Scripting API: security metadata in API

### ISSUE THREAD

**Issue by ereshetova** [w3c/wot-scripting-api#91]
Some WoT Scripting API methods need to have a way to provide security-related parameters for these security-related use cases:

1. On server side defining access control for actions & properties: When building a new (or updating) ThingDescription object and ExposedThing, there has to be a way for a script to specify security-related information that client would need to have in order to be authorized to do certain action or access a property for reading or writing (calling the invokeAction(), the writeProperty() and the readProperty() methods). 

2. On server side defining access control for notifications & Object Changes: When building a new (or updating) ThingDescription object and ExposedThing, there has to be a way for a script to specify security-related information that client would need to have in order to be authorized to subscribe to various events on Thing, property change or TD change (being able to get callbacks from the onEvent(), the onPropertyChange(), and the onTDChange() methods)

3. On client side providing security-metadata for above two cases: after the client reads the TD and knows what credentials it needs to invoke certain method or subscribe for certain info, it goes and obtains the required credentials (if doesn't have), and then it needs to be able to supply them somehow to the server. It might be supplying an authorization token or certificate or anything else, but the information needs to be delivered.  For example, when invoking writeProperty() or onEvent() methods.

**Comment by ereshetova** [comment]
For your review: @zolkis, @mmccool

**Comment by zolkis** [comment]
Currently we can add semantic annotations to the Thing object. Security metadata could be one of these. However, it is under discussion whether the so-called "TD metadata" (.name, .base, .security, and, well, .interaction) should be exposed directly.

**Comment by mkovatsc** [comment]
Just like the bindings, security depends on the WoT Runtime implementation. A script should not be able to set this directly; it is added vy the runtime based on the available security mechanisms and the security metadata passed along with, but separated from the script -- or attached to the runtime a script is instantiated in.

**Comment by zolkis** [comment]
But the TD has security metadata, and scripts can create TDs. Therefore scripts should be able to add security metadata (this is the core of the issue). 

If you think TD's created by scripts should be incomplete (without security metadata), then I question the usefulness of creating TDs by scripts. 

The ExposedThing API would be a lot simpler if the TD would be supplied as a blob, created by some tool or external library (also containing security metadata) and the only thing to be added by scripts would be the handler code (if differing from the default handlers).

**Comment by mkovatsc** [comment]
What concrete security information would you then add via script and from where would you get it?

**Comment by zolkis** [comment]
You listed some in the IG current practices: auth type, tokens, ACLs, etc.

To be clear, my personal opinion is that security provisioning should be part of provisioning (which sadly is out of scope for the WG), and scripts using the WoT object should not provision security. This should be on a different API object (that is used for provisioning). 

So I would rule this issue as out of scope (also from the TD), but I understand the issue and try to cooperate with experimental support until we figure out how to create full stack WoT solutions and where to best delegate the issue.

**Comment by mkovatsc** [comment]
So how would a script know what security mechanisms the runtime system provides? Read it through the API and then feed it back again through the API?

Neither should a script manage what IP addresses the runtime host has and fill it into the TD. Thus, it is just natural that these information are filled in by the runtime system (available bindings, socket address, and security metadata). Only the application-specific parts need to be constructed through the application script.

**Comment by mkovatsc** [comment]
Also note that the TD feature to describe security metadata is independent from the Scripting API. Things can be implemented however the creator chooses, while the resulting network-facing API is described by a TD. This network-facing API description needs to include security metadata to be a complete and useful description.

**Comment by zolkis** [comment]
But when there are choices supported by the runtime, the script writer may want to specify one or more of them.

**Comment by mkovatsc** [comment]
That might be not good for script re-usability, no?

Wouldn't it be better to do this configuration in the same process that installs a script in  a specific runtime etc.?

**Comment by zolkis** [comment]
> That might be not good for script re-usability, no?

Client (consumed Things) code could be reusable, but here we are talking about scripts that expose Things. These need not be fully reusable. Besides, the non-reusable parts could be separated. Then, since script deployment is out of WG scope, the script reusability argument doesn't have ground anyway (it's a non-usecase).

> Wouldn't it be better to do this configuration in the same process that installs a script in a specific runtime etc.?

Yes, it would, for provisioning. I would separate the provisioning + configuration process to a totally different setup and API mechanism (needs bootstrapping, trusted connection, etc).

However, in an already provisioned system (e.g. an existing OCF network with a WoT bridge), a WoT script developer would want to e.g. add ACLs to the newly created Thing that is translated by protocol bindings to OCF ACL provisioning (the WoT Thing is a client in the OCF network). 

Identity can be translated by the runtime, security mechanisms can be encapsulated, but developer/user choices bubble up to the API. 

Since this is a non-trivial problem, I understand the wish to defer this to another API/mechanism. But then we can't create a Thing from Scripting alone - which may be fine.

Sadly, real problems like provisioning, script deployment etc unfortunately are out of the WG scope (but are real showstoppers if not solved).

**Comment by zolkis** [comment]
I will do this by next Wednesday (edit this comment into it).

**Comment by ereshetova** [comment]
I think I will also wait for the OCF use case to comment further. It is always better to have a real example to discuss.

**Comment by zolkis** [comment]
In OCF, a device must be _owned_. A _device_ contains multiple _resources_ that must have an _identity_ and must have _access control list_ (access control is per resource and not per property). All security metadata is stored at device level in a special resources. So access right must be obtained by clients prior to making requests for a resource, IOW even  a client should be provisioned with the device that hosts the resource, either directly, or through a proxy service.

First, a device must undergo _onboarding_ (establishing ownership and trust chain, based on several methods to share a secret through a separate trusted channel, or based on manufacturer certificate). Then the device is _provisioned_ with credentials and ACLs, then _configured_ (name, location, language, regional settings etc). 

For every new client, the server device should be provisioned with an ACL for the client, and the client device should be provisioned with the access credentials, usually via a _credential management service_ (CMS) run on behalf the device owner. 

A WoT implementation (as OCF client) should comply to the provisioning process:
- Discover OIC/OCF devices/resources, introspect resources.
- Figure out the provisioning service/CMS URL and other device info.
- Claim client role, permission, validity etc for given access.
- Make requests to resources.

In my current understanding, a WoT implementation should use a module for OCF interactions as an OCF client loaded on demand by protocol bindings.

@mmccool, @mjkoster please check.

**Comment by zolkis** [comment]
As for exposing a WoT Thing as an OCF server, the device containing the WoT Thing should be provisioned as an OCF device, i.e. should be owned, provisioned, configured, and implement the resources it exposes. It should also undergo OCF certification.

**Comment by mmccool** [comment]
Regarding use cases and what security metadata should be provided... as it happens I have been reviewing the security metadata provided for in OpenAPI 3.0.   OCF uses OpenAPI 2.0 currently for representing its metadata so we can also look at how OCF represents its security metadata in OpenAPI.  

I am strongly thinking of recommending (we can discuss in the Security TF call today) that we simply (as a starting point at least) adopt the OpenAPI 3.0 metadata model for security metadata.  For one thing, they already cover all the various ways in which web APIs are authenticated and secured, so that will cover the HTTP use cases for WoT.  If we also augment that with the OCF security metadata as expressed in OpenAPI, we'll have covered a significant majority of use cases... at least for resource-oriented APIs.   There may be some more specific CoAP use cases not included in OCF but we can add those incrementally.

As Zoltan mentions though, one key issue is that OpenAPI (and OCF) is organized per-resource, not per-interaction (eg, it lists the resources, then for each resource, the interactions; WoT lists the interactions, then for each interaction the resource used).  This might cause some issues too.   However, as a larger issue currently security metadata is per-Thing, which is probably too coarse a granularity.   One other related thing that OpenAPI does that I *know* will cause problems with JSON-LD is that it allows security metadata to be specified at the top level, then be overridden for each resource if a particular resource uses a different setup.   We will have problems expressing this pattern in JSON-LD for the same reason we have problems with default values.   However, I do think we *probably* need some finer-grained way to specify security per interaction.   Any ideas?   Maybe we specify a set of named "configurations" at the top level and then apply them to interactions?

Regarding the scripting API: right now we say "security provisioning is out of scope" but maybe this is taking too narrow a view.  We did this initially for convenience (to keep the scope of the TD and Scripting API narrow) but if it's causing problems or will result in an unusable spec we should revisit it.  In particular we may want to have multiple API objects for use at different points in a Thing's lifecycle.  Also, although a script can't specify what mechanisms a platform makes available, I think certainly the *list* of security mechanisms supported by a runtime should be available for introspection to a script, and it should be possible to programmatically provide the necessary parameters to those mechanisms.   It's certainly reasonable to say this interaction and parameter-setting happens at a different phase than the "operational" phase.   But I think we already do that: when a Thing is being set up (when we are specifying all its interactions and have not exposed it yet), it's obviously not operational yet: it only becomes operational after we expose it.

**Comment by mmccool** [comment]
PS I need to look again at how exactly OCF handles security metadata.  It may not all be in the OpenAPI introspection object, for instance, since OCF provides several way to introspect, and security in particular has a special set of resources assigned to it.

**Comment by mmccool** [comment]
I just reviewed the OCF spec (and the Iotivity implementation) and indeed security is generally managed by manipulating a special set of resources.  Even provisioning is handled this way: credentials are provisioned by POSTing them to /oic/sec/cred (see https://wiki.iotivity.org/iotivity_security_architecture_overview).   This corresponds, I guess, to having a "Management Thing" but is quite different from how we have been thinking about security in WoT.

I should mention when I did my metadata bridge from OCF to WoT I actually read certain resources, like oic.res, and translated them into entries in the TD.   It would be possible to translate information that OCF keeps in resources into values in the TD.   Other values that are in resources in OCF but that we might want to put in the TD of a WoT interface are the UUID and the ACL.   Part of the issue I think is that OCF has tended to put static metadata in RAML/Swagger, and per-device metadata in resources (mostly...).  In WoT we take a "prototype" class approach and put both class (static) and instance (variable) metadata in the same description.

**Comment by mkovatsc** [comment]
> As Zoltan mentions though, one key issue is that OpenAPI (and OCF) is organized per-resource, not per-interaction (eg, it lists the resources, then for each resource, the interactions; WoT lists the interactions, then for each interaction the resource used).

Our Interactions (usually) correspond to OCF resources; the OCF resource properties are the elements of a complex WoT Property. The fact that WoT can have multiple bindings should not affect the authorization needed for the Interaction/resource.

> a larger issue currently security metadata is per-Thing, which is probably too coarse a granularity

According to Zoltan, "**the device is provisioned** with credentials and ACLs" and "the **client device should be provisioned** with the access credentials" (i.e., similar to Things level). If you compare to OAuth, it is the same: the a client has a token/credentials; this token can have different audiences/scopes, allowing for a more fine-grained authorization (e.g., Interactions or maybe Capabilities).

> [OpenAPI] allows security metadata to be specified at the top level, then be overridden for each resource if a particular resource uses a different setup

We could think about allowing Interaction-level security metadata. Another way would be to question if this should then still be the same "Thing", if different security contexts are required.

> we would have to "consume" an OCF device represented as a WoT thing incrementally

No, the idea is that a TD is generated with all the required metadata in place. If a device offers some metadata through a resource/API in some other ecosystem, whatever generates the TD takes this metadata from there and puts it directly into the TD.

> [In OCF] security is generally managed by manipulating a special set of resources. Even provisioning is handled this way: credentials are provisioned by POSTing them to /oic/sec/cred 

This fits how I have been thinking about security in WoT. I doubt (and cannot remember) that that there is code in the Resource implementations of IoTivity that sets ACLs or provisions credentials directly; at best it can introspect its ACLs or credentials to enforce security. In WoT, enforcement should take place in the runtime system, not the scripts, so we would not need this security introspection.

> WoT we take a "prototype" class approach and put both class (static) and instance (variable) metadata in the same description

A TD is always **instance-specific**. It has metadata for a specific Thing instance.

**Comment by ereshetova** [comment]
There are a number of discussions going here now, so I will try to reply separately to them. 
So first about OCF client case that Zoltan described:

>A WoT implementation (as OCF client) should comply to the provisioning process:

>Discover OIC/OCF devices/resources, introspect resources.
>Figure out the provisioning service/CMS URL and other device info.
>Claim client role, permission, validity etc for given access.
>Make requests to resources.

This is also in line with my understanding on OCF client behavoir, so we are on the same page here. 

>In my current understanding, a WoT implementation should use a module for OCF interactions as an >OCF client loaded on demand by protocol bindings. See https://wiki.iotivity.org/provisioning

OK, so I guess here we are saying that OCF protocol binding module would do all the necessary above steps here, which is fine, but I do not fully understand how it would look like from WoT level perspective. 
Would it be possible to see how it would look like in terms of code (pseudocode)? Smth like:

1. Wot client script makes a broadcast for available devices around using discover() method.
2. One of the returned result is a TD that says that there is an "OCF device" around. Wot client script wants to talk to this device now and use its resources. 
3.  Wot script performs some magic (what is the API for this or how does it happen?)  to activate OCF protocol binding and let it do its own magic for doing the steps like "Figure out the provisioning service/CMS URL and other device info, Claim client role, permission, validity etc for given access."
4. After step 3 completes successfully, a Wot client script can just call methods/resources from TD produced from this OCF device and everything wither works smoothly.
5. Everything still works smoothly until the credentials requested in 3 expire (or were revoked in CMS server or anything else happen). What happens in this case? I guess it is again the OCF protocol binding that should handle obtaining new credentials. How will it do it? By noticing that resources requests that WoT client script sends out come back with security error? If so, then it doesn't return the error up to the WoT level but first try to get new ones, and only after it fails, then it returns a security error back to WoT client script saying "I did everything I could but still no luck with access".

I think we really need a sequence flow diagram here as we discussed in security call, because otherwise there are so many details that we can forget....

**Comment by ereshetova** [comment]
>As for exposing a WoT Thing as an OCF server, the device containing the WoT Thing should be >provisioned as an OCF device, i.e. should be owned, provisioned, configured, and implement the >resources it exposes. It should also undergo OCF certification.

This is actually quite a complex case. Do we really support it in WoT? Do we expect WoT scripts to do create OCF-compliant servers? I guess not in run-time, right? So maybe we don't have to worry about it for scripting API?

**Comment by mkovatsc** [comment]
I assume all this can be done with an OCF-compliant *runtime system* that can also produce TDs, whereas the script only provides the resource implementations. So far I have not seen a real blocking point for that.

**Comment by mkovatsc** [comment]
0. An Runtime System / Servient is set up with an already enabled OCF binding -- or rather with a CoAP binding -- and the runtime for the script is configured with the provisioning 

1. The script finds an OCF device through discover().

2. When the ConsumedThing is instanciated, the runtime Claim client role, permission, validity etc for given access

3. The script does nothing as the runtime system has both the TD and provisioning service.

4. Just use the Thing Interactions.

5. Runtime system should reprovision/reclaim or decommission client if script is done.

**Comment by mmccool** [comment]
I just created a pull request with a strawman for how I think security metadata should be handled in a TD.  See here: https://github.com/w3c/wot-security/pull/74

This is just a strawman, and has tons of issues.  I also need to go back and look at a number of other comments, including Matthias' example, and try to fold them in.  But the whole idea of a strawman is to get something out there and iterate ;)

A few random comments on the discussion above:
- OCF security is complicated but we don't necessarily need to include all that complexity in the TD.  The *minimum* metadata is just a declaration that the device in question uses the OCF security model (otherwise it is indistinguishable from any other coap device).  However, the assumption is that the underlying OCF-aware runtime system, once it knows it is talking to an OCF device, will do all the neccessary introspection, setup, etc. automatically.   I think this is what Zoltan was getting at: for many devices we don't want to give ALL the metadata, we just want to give enough to bootstrap the connection process, and let existing mechanisms be used to handle the rest.  As another example, we don't specify certificates, etc. since TLS already does that as part of the connection process.
- My suggested approach includes the ability to have multiple security configurations and apply them to different interactions in the same thing.  While we *could* break up Things into Sub-Things with uniform security I think this will be messy.   My model also allows different interactions to use different mechanisms or levels of security, and also alllows interactions to have multiple security mechanisms, or different mechanisms for different protocols (eg http vs coap) within the same interaction.  It also allows security configurations to be reused rather than being repeated everwhere.  Some of these ideas may or may not work in JSON-LD, though.  If necessary we can ditch the reuse mechanism without losing too much.
- I looked at OpenAPI for a collection of web API access control mechanisms.  Some of these (eg OpenID) may or may not make sense here.
- I think we should discuss the "instance/class" issue I mentioned above further.  People need a way to talk about classes of devices too, and the easiest way to do that is via a "prototype".  Even if you tell them not to... they will.   By "prototype" I mean specifically an "example" thing people clone and modify to represent a specific device.  This is not that directly related to security.

**Comment by zolkis** [comment]
> Our Interactions (usually) correspond to OCF resources; the OCF resource properties are the elements of a complex WoT Property. 

This needs some rethinking, and it is very important in mapping security metadata. In OCF, a resource is constituted from all properties making up a _resource representation_ that can be bound to one security policy, as recommended by the OCF Security spec. So if OCF resources respect that security design (IIUR), there will be a resource if we want to get information, and another resource if we want to write information. For instance, an action (translated to writing a certain kind of property) will be on a different resource than reading state. This would be in line with Matthias' statement. However, I would not state a rule that _in general_, WoT interactions correspond to OCF resources. An OCF resource could map to a WoT Thing if all properties are with one security policy; that means, multiple Property read interactions, or multiple Property write + action interactions. So there are cases when a single OCF resource corresponds to multiple WoT interactions.

WoT is actually a simplified model compared to OCF: in OCF there is _platform_ (the HW), _device_ (the SW stack), _resource_ (Thing or Interaction). In OCF, one can separately discover platforms, devices and resources. The business logic of the application will tell what and in what order needs to be discovered. That means WoT cannot support all scenarios possible with OCF objects, just a subset, like discovering resources, then for each desired resource, get platform and device information, then create a ConsumedThing based on that.

When a script discovers one or more OCF resources, the WoT implementation will likely have to make a separate ConsumedThing from every OCF resource. So actually we don't need to attempt mapping WoT and OCF concepts too exactly: theoretical mapping is hard, practical mapping is easier (at least for ConsumedThing/OCF client).

**Comment by zolkis** [comment]
> OCF security is complicated but we don't necessarily need to include all that complexity in the TD. The minimum metadata is just a declaration that the device in question uses the OCF security model (otherwise it is indistinguishable from any other coap device). However, the assumption is that the underlying OCF-aware runtime system, once it knows it is talking to an OCF device, will do all the neccessary introspection, setup, etc. automatically. I think this is what Zoltan was getting at: for many devices we don't want to give ALL the metadata, we just want to give enough to bootstrap the connection process, and let existing mechanisms be used to handle the rest. 

That is right. However, the _minimum_ information will surely be more than just "this is an OCF device". We need some kind of "minimal metadata" representation for OCF, and likely for other protocol suites as well.

**Comment by mmccool** [comment]
Zoltan Re minimum metadata for OCF: you're probably right.  I guess I should have said "at a minimum (eg at least) we have to identify it as an OCF device".  My point is though that we should not just throw metadata into the WoT TD because we have it.  Instead we need to figure out the minimum necessary to bootstrap security on the target, and not repeat information available be other means.  I think the way to do that is hypothesize a minimum and try to get it to work, then iterate, adding only what is absolutely necessary.

Regarding your example of "separate resources for reading and writing" this reminded me of a related issue: we may require different credentials for read access vs read-write access.  For instance, for some property we may decide to make reading open but writing require an API key.   How can we represent that?   Multiple protocol bindings restricted to different methods (eg http:get and http:post) and with different security methods?  That would work, and also allows for different resources (or even protocols!) to be used for reading and writing a single conceptual property.

Matthias, regarding my comment about "interaction-first vs resources-first organization": they are the same if resources and interactions map one-to-one.  You could argue that in a good RESTful API design this would be the case.  Maybe... but not necessarily.   Another example actually would be the use of hybrid schemes like MQTT messages to handle "observe" in an otherwise HTTP-based API.   The MQTT resources might use a completely different URL scheme to represent topics related to properties.  In this case one interaction maps to (refers to) several resources.   There are probably converse examples where a single resource is referenced from multiple interactions...

**Comment by mmccool** [comment]
Anyway, please take a look at the PR and make specific comments there. I'm intending to beef that PR up over the next few days (working through a number if different authentication and authorization schemes, OCF, etc), making lists of issues, and bringing the whole mess to the table for discussion in the next Security TF.

**Comment by mkovatsc** [comment]
@zolkis For the F2F session, could you make the discussion here "digestable"? I think it is important to look at concrete use cases where security metadata would need to be passed from the script to the runtime or vice-versa.
